src/herder/HerderImpl.cpp

```diff
@@ -1086,55 +1086,66 @@ HerderImpl::restoreSCPState()
-    decoder::decode_b64(latest64, buffer);
+        std::vector<uint8_t> buffer;
+        decoder::decode_b64(state, buffer);
 
-    xdr::xvector<SCPEnvelope> latestEnvs;
-    xdr::xvector<TransactionSet> latestTxSets;
-    xdr::xvector<SCPQuorumSet> latestQSets;
-
-    try
-    {
-        xdr::xdr_from_opaque(buffer, latestEnvs, latestTxSets, latestQSets);
+        xdr::xvector<SCPEnvelope> latestEnvs;
+        xdr::xvector<TransactionSet> latestTxSets;
+        xdr::xvector<SCPQuorumSet> latestQSets;
 
-        for (auto const& txset : latestTxSets)
-        {
-            TxSetFramePtr cur =
-                make_shared<TxSetFrame>(mApp.getNetworkID(), txset);
-            Hash h = cur->getContentsHash();
-            mPendingEnvelopes.addTxSet(h, 0, cur);
-        }
-        for (auto const& qset : latestQSets)
-        {
-            Hash hash = sha256(xdr::xdr_to_opaque(qset));
-            mPendingEnvelopes.addSCPQuorumSet(hash, qset);
-        }
-        for (auto const& e : latestEnvs)
+        try
         {
-            getSCP().setStateFromEnvelope(e.statement.slotIndex, e);
-        }
+            xdr::xdr_from_opaque(buffer, latestEnvs, latestTxSets, latestQSets);
 
-        if (latestEnvs.size() != 0)
+            for (auto const& txset : latestTxSets)
+            {
+                TxSetFramePtr cur =
+                    make_shared<TxSetFrame>(mApp.getNetworkID(), txset);
+                Hash h = cur->getContentsHash();
+                mPendingEnvelopes.addTxSet(h, 0, cur);
+            }
+            for (auto const& qset : latestQSets)
+            {
+                Hash hash = sha256(xdr::xdr_to_opaque(qset));
+                mPendingEnvelopes.addSCPQuorumSet(hash, qset);
+            }
+            for (auto const& e : latestEnvs)
+            {
+                // Only restore all envelopes for the latest slot
+                if (e.statement.slotIndex == lcl.header.ledgerSeq ||
+                    e.statement.pledges.type() == SCP_ST_EXTERNALIZE)
+                {
+                    getSCP().setStateFromEnvelope(e.statement.slotIndex, e);
+                    mLastSlotSaved =
+                        std::max<uint64>(mLastSlotSaved, e.statement.slotIndex);
+                    broadcast = broadcast ||
+                                e.statement.slotIndex == lcl.header.ledgerSeq;
+                }
+            }
+        }
+        catch (std::exception& e)
         {
-            mLastSlotSaved = latestEnvs.back().statement.slotIndex;
-            startRebroadcastTimer();
+            // we may have exceptions when upgrading the protocol
+            // this should be the only time we get exceptions decoding old
+            // messages.
+            CLOG(INFO, "Herder") << "Error while restoring old scp messages, "
+                                    "proceeding without them : "
+                                 << e.what();
         }
     }
-    catch (std::exception& e)
+
+    if (broadcast)
```

Comment:
correction on broadcast flag: it's probably better to always call that function, I am not sure why it was behind that condition

Reply:
Updated.